KAFKA-6795: Added unit tests for ReplicaAlterLogDirsThread

apovzner opened the pull request:
Added unit tests for ReplicaAlterLogDirsThread. Mostly focused on unit tests for truncating logic.

Fixed  ReplicaAlterLogDirsThread.buildLeaderEpochRequest() to use future replica's latest epoch (not the latest epoch of replica it is fetching from). This follows the logic that offset for leader epoch request should be based on leader epoch of the follower (in this case it's the future local replica). 

Also fixed PartitionFetchState constructor that takes offset and delay. The code ignored the delay parameter and used 0 for the delay. This constructor is used only by another constructor which passes delay = 0, which luckily works. 

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

apovzner committed:
KAFKA-6795: Added unit tests for ReplicaAlterLogDirsThread; and fixed one PartitionFetchState constructor

lindong28 commented on core/src/test/scala/unit/kafka/server/ReplicaAlterLogDirsThreadTest.scala:
```diff
@@ -0,0 +1,510 @@
+    val responseCallback: Capture[Seq[(TopicPartition, FetchPartitionData)] => Unit]  = EasyMock.newCapture()
+
+    val leaderEpoch = 2
+    val futureReplicaLEO = 200
```
nits: would it be slightly more realistic to let futureReplicaLEO be smaller than he LEO of the current replica?

apovzner committed:
buildLeaderEpochRequest should use future replica's latest epoch

lindong28 commented on core/src/test/scala/unit/kafka/server/ReplicaAlterLogDirsThreadTest.scala:
```diff
@@ -0,0 +1,525 @@
+    // next time we loop, getReplicaOrException will return replica
+    thread.doWork()
+
+    //Now the final call should have actually done a truncation (to offset 156)
```
Is the number `156` in the comment outdated?

lindong28 commented on core/src/test/scala/unit/kafka/server/ReplicaAlterLogDirsThreadTest.scala:
```diff
@@ -0,0 +1,525 @@
+    val responseCallback: Capture[Seq[(TopicPartition, FetchPartitionData)] => Unit]  = EasyMock.newCapture()
+
+    val leaderEpoch = 5
+    val futureReplicaLEO = 200
```
nits: would it be slightly more realistic to let futureReplicaLEO be smaller than he LEO of the current replica? Normally futureReplicaLEO should be <= replicaLEO.

lindong28 reviewed (commented):
Thanks for the patch! Looks good overall. Just left some minor comments.

apovzner committed:
addressed review comments

apovzner commented:
Hi @lindong28, thanks for your review. I addressed your comments. I left one test case where future replica's LEO is bigger than replica's LEO so we test all the paths.

apovzner commented:
cc @junrao

lindong28 commented:
Thanks @apovzner for the update. The patch LGTM. We can wait for Jun's comment.

apovzner commented:
Two unrelated test failures:
15:58:46 kafka.server.DynamicBrokerReconfigurationTest > testThreadPoolResize FAILED
https://issues.apache.org/jira/browse/KAFKA-6823

17:20:49 kafka.server.DynamicBrokerReconfigurationTest > testAddRemoveSslListener FAILED
https://issues.apache.org/jira/browse/KAFKA-6824

apovzner commented:
retest this please

apovzner commented:
Again failures in kafka.server.DynamicBrokerReconfigurationTest > testAddRemoveSslListener. I checked PR builds of other PRs and the test also fails quite frequently on other PR builds.

apovzner commented:
retest this please

lindong28 commented:
@apovzner Since this is a known test failure in the trunk, I think we probably don't need to block this PR on this test failure.

lindong28 commented:
@apovzner Would you like to wait for Jun's review, or do you want this patch to be committed? Thanks!

lindong28 commented:
@apovzner BTW, I am happy to get the first few rounds of review for KIP-279.

apovzner commented:
@lindong28 Could you please commit this patch? I just wanted Jun to be cc'd. And then I will rebase my KIP-279 PR and add few more related tests. Would be great if you could review KIP-279 PR! Let me first do few more updates (today) and I will let you know when ready.

lindong28 commented:
@apovzner Sure thing! The PR has been merged.
